### ISSUE TITLE
ImageDataGenerator producing incorrect batch sizes

### ISSUE THREAD

**Issue by jacobkimmel** [fchollet/keras#7299]
In Keras 2.0's ImageDataGenerator, I'm receiving an error where the generator returns an incorrect batch size after a periodic number of epochs. 

**Relevant versions:**
Keras-2.0.6 (installed from Github, also tried downgrading to 2.0.0 & still saw the issue).
TensorFlow-1.2.1

Training data is arranged as `.png` in three class folders, with the following number of images in each per class directory `ls | wc -l`:

```
00_class0   969
01_class1   1423
02_class2   335
```

Using a batch_size of 64, the generator returns a batch of 39 every 43 batches.

It seems like the iterator is just returning the "remaining"  samples at the 43rd epoch, rather than shuffling the index_array and starting over?

Gist to reproduce the bug:
https://gist.github.com/jacobkimmel/1f56d01187c6ca11ee716b6f224ffac6

fwiw, I've used Keras extensively since 1.2 and have never encountered an issue like this. 

Any ideas?

Thank you!

- [X] Check that you are up-to-date with the master branch of Keras. You can update with:
pip install git+git://github.com/fchollet/keras.git --upgrade --no-deps

- [X] If running on TensorFlow, check that you are up-to-date with the latest version. The installation instructions can be found [here](https://www.tensorflow.org/get_started/os_setup).

- [NA] If running on Theano, check that you are up-to-date with the master branch of Theano. You can update with:
pip install git+git://github.com/Theano/Theano.git --upgrade --no-deps

- [X] Provide a link to a GitHub Gist of a Python script that can reproduce your issue (or just copy the script here if it is short).
